### PR TITLE
Create dim dates

### PR DESCRIPTION
**Pull request by rosie-t**
Linked to [this ticket](https://github.com/orgs/cbfacademy/projects/46?pane=issue&itemId=171853389&issue=cbfacademy%7Cfittrack_analytics%7C26)

Creating a dim_date table with one row per day. This table isn't macro-dependent (as it was previously.
Calendar attributes are calculated using SQL and a weekend boolean value is added. 
The idea is that this dim_table can be joined where time analysis is needed. 

Tests are also added and pass.

### CODE REVIEW

**Review by RJO23 — APPROVED** [review]
Looks good to me!

File: models/marts/dim_date.sql [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+    format_date('%A', date_day) as day_of_week_name,
+    case
+        when extract(dayofweek from date_day) in (1, 7) then true
+        else false
+    end as is_weekend,
+    extract(day from date_day) as day_of_month,
+    extract(dayofyear from date_day) as day_of_year,
+
+## Week information ##
+
```
Comment (harjeetkalsi) [inline models/marts/dim_date.sql]: This looks good overall. One thing I’d flag is that we are mixing ISO week logic with calendar year logic here.

`week_of_year` uses `extract(isoweek ...)`, but `year_number` uses the calendar year. Around year boundaries that can produce confusing combinations, for example ISO week 1 paired with the previous calendar year, or vice versa.

I think we should either add an `iso_year` field to keep the week logic consistent, or switch `week_of_year` to a non-ISO definition if we want this dimension to stay purely calendar-based.

File: models/marts/dim_date.sql [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+}}
+
+## What dates to start and end from, can be adjusted easily ##
+
+with parameters as (
+
+    select
+        date('2023-01-01') as start_date,
+        date('2050-12-31') as end_date
+
```
Comment (harjeetkalsi) [inline models/marts/dim_date.sql:14]: This is not a blocker, but it would be better if start_date and end_date were configurable via var() rather than fixed in the SQL.

That makes the model easier to reuse and avoids editing code just to extend the range later.

Example idea:

date('{{ var("dim_date_start", "2023-01-01") }}') as start_date,
date('{{ var("dim_date_end", "2050-12-31") }}') as end_date

File: models/marts/dim_date.sql [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+    select
+        date_day
+    from parameters,
+    unnest(generate_date_array(start_date, end_date)) as date_day
+
+)
+
+select
+
```
Comment (harjeetkalsi) [inline models/marts/dim_date.sql:31]: A lot of date dimensions benefit from a key like 20260115 for joining and reporting convenience.

Not always required, but worth considering if the rest of the project is moving toward marts and dimensions.

File: models/marts/dim_date.yml [hunk 4]
```diff
@@ -0,0 +1,118 @@
+version: 2
+
```
Comment (harjeetkalsi) [inline models/marts/dim_date.yml:2]: The column tests are good, but this model would benefit from checking specific tricky dates such as:

leap day, e.g. 2024-02-29
year boundary, e.g. 2023-12-31 and 2024-01-01
month end and quarter end behavior

Even lightweight unit tests for a few known dates would give extra confidence.

**Review by harjeetkalsi — CHANGES_REQUESTED** [review]
Nice addition overall. The model is clean and useful, and the core calendar fields look good.

The main thing I’d like us to address before merging is the week logic. We currently use `isoweek` for `week_of_year`, but calendar year for `year_number`, which can give confusing results around year boundaries. I think we should either add `iso_year` or make the week definition fully non-ISO for consistency.

A couple of smaller follow-ups:

* worth making the start/end dates configurable via `var()`
* optional, but a `date_key` could be useful if this is going to be a commonly joined dimension

**Comment by rosie-t** [comment]
Changes since last PR 
- Added in better ISO logic and renaming of columns
- Added testing to account for this

**Review by harjeetkalsi — APPROVED** [review]
This is a really strong PR, well done Rosie! 

You’ve taken the feedback on board and implemented the configurable date range cleanly using `var()`, which makes the model much more flexible across environments (by that I meean: test, dev prod). The use of `generate_date_array` to build the date spine is also a great choice. 

The overall structure of the `dim_date` model is clear, well organised, and easy to follow, with good coverage across day, ISO week, month, quarter, and year.

Quality of your tests is great too. You’ve gone beyond basic checks and are validating logic and edge cases. The `is_weekend` test enforces consistency nicely, and the ISO year/week tests are strong as they cover boundary cases too. 

You've built something that is reliable, testable, and prod ready.

Happy to approve  :)
